### PR TITLE
resolves #1 add typing of boticot and messenger responses

### LINKED ISSUE #1: Add typing of boticot api response

### PR DESCRIPTION
**Pull request by leooel88**
Add typings boticot and messenger typings

### CODE REVIEW

File: src/client/nlu.ts [hunk 1]
```diff
@@ -1,9 +1,13 @@
 import axios from 'axios';
 import * as correlator from 'express-correlation-id';
+import { BoticotData } from '../../typings/global';
 
-export const parseNlu = async (input: string, userId: String): Promise<any> => {
+const tunnel = require('tunnel'); //
```
Comment (omar-mellouli) [inline src/client/nlu.ts]: to remove

File: package.json [hunk 2]
```diff
@@ -13,7 +13,9 @@
     "@types/node": "^14.0.11",
     "express": "^4.17.1",
     "express-correlation-id": "^2.0.1",
-    "axios": "^0.21.1",
+    "axios": "^0.27.2",
+    "dotenv": "^16.0.2",
+    "tunnel": "^0.0.6",
```
Comment (omar-mellouli) [inline package.json]: to remove

File: src/client/nlu.ts [hunk 3]
```diff
@@ -1,9 +1,13 @@
...
+
+export const parseNlu = async (input: string, userId: string): Promise<BoticotData> => {
     const { API_URL, AGENT_NAME } = process.env;
     const url = `${API_URL}/nlu/agents/${AGENT_NAME}/parse`;
+
```
Comment (omar-mellouli) [inline src/client/nlu.ts]: to remove

File: src/express.ts [hunk 4]
```diff
@@ -4,27 +4,29 @@ import routes from './routes';
 import * as correlator from 'express-correlation-id';
 
 class App {
-    public appExpress: Application;
```
Comment (omar-mellouli) [inline src/express.ts:7]: rollback refactoring

File: typings/global.d.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+export interface MessagingEventMessage extends MessagingEventBasic {
+	message: Message
+	postback?: never
+}
+
+export interface MessagingEventPostback extends MessagingEventBasic {
+	message?: never
+	postback: Webhook // BIZARRE
```
Comment (omar-mellouli) [inline typings/global.d.ts]: remove comment

File: typings/global.d.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+	text: string
+	quick_reply?: never
+}
+
+export interface MessageQuickReplies {
+	quick_reply: QuickReply
+	text?: never
+}
+
+export class ImageLoadResponse {
```
Comment (omar-mellouli) [inline typings/global.d.ts]: to remove

File: typings/global.d.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+	data: ImageLoadData
+}
+
+export class ImageLoadData {
```
Comment (omar-mellouli) [inline typings/global.d.ts]: to remove

File: typings/express.d.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+declare namespace Express {
```
Comment (omar-mellouli) [inline typings/express.d.ts]: check if mandatory

File: tsconfig.json [hunk 9]
```diff
@@ -10,6 +10,7 @@
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true,
     "target": "es2017",
-    "lib": ["es2017"]
+    "lib": ["es2017"],
+    "skipLibCheck": true,
```
Comment (omar-mellouli) [inline tsconfig.json:14]: check if needed

File: src/utils/messenger.ts [hunk 10]
```diff
@@ -1,6 +1,18 @@
+import {
+	Suggestion,
+	QuickReply,
+	Link,
+	MessengerTextAndQuickReplyRequest,
+	MessengerTextUrlAndQuickReplyRequest,
+	MessengerTextRequest,
+	MessengerTextImageAndSuggestion,
```
Comment (omar-mellouli) [inline src/utils/messenger.ts]: to remove

File: src/services/messenger.ts [hunk 11]
```diff
@@ -65,60 +81,72 @@ class MessengerService {
...
+                return messageText(sender_psid, boticot_data.response.fulfillment_text)
             }
         } else {
             throw Error('Missed Fulfillment Text in response')
         }
     }
 
-    async handleQuickReplies(sender_psid, received_postback) {
-        let payload = received_postback.quick_reply.payload;
+    async handleQuickReplies(sender_psid: string, received_message: Message) {
```
Comment (omar-mellouli) [inline src/services/messenger.ts]: rollback

File: src/services/messenger.ts [hunk 12]
```diff
@@ -1,17 +1,31 @@
+import { Response } from 'express'
+
```
Comment (omar-mellouli) [inline src/services/messenger.ts]: remove

File: typings/global.d.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+	message: MessengerMessage
+}
+
+export type MessengerRequest =
+	| MessengerTextRequest
+	| MessengerTextAndQuickReplyRequest
+	| MessengerTextUrlAndQuickReplyRequest
+	| MessengerTextImageAndSuggestion
+
+export class Recipient {
```
Comment (omar-mellouli) [inline typings/global.d.ts]: remove class (for all)

File: typings/global.d.ts [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+export interface MessengerTextAndQuickReplyRequest extends MessengerRequestBasic {
+	messaging_type: string
+	message: MessengerMessage
+}
+
+export interface MessengerTextUrlAndQuickReplyRequest extends MessengerRequestBasic {
+	message: MessengerMessage
+}
+
+export interface MessengerTextImageAndSuggestion extends MessengerRequestBasic {
```
Comment (omar-mellouli) [inline typings/global.d.ts]: remove duplicated

File: environment.d.ts [hunk 15]
```diff
@@ -0,0 +1,15 @@
+declare global {
```
Comment (omar-mellouli) [inline environment.d.ts]: to remove

File: package.json [hunk 16]
```diff
@@ -13,7 +13,9 @@
     "@types/node": "^14.0.11",
     "express": "^4.17.1",
     "express-correlation-id": "^2.0.1",
-    "axios": "^0.21.1",
+    "axios": "^0.27.2",
+    "dotenv": "^16.0.2",
```
Comment (omar-mellouli) [inline package.json]: to remove

**Review by omar-mellouli — CHANGES_REQUESTED** [review]

**Comment by omar-mellouli** [comment]
rename commit message

**Comment by omar-mellouli** [comment]
please squash commit

File: src/server.ts [hunk 17]
```diff
@@ -1,3 +1,5 @@
+import 'dotenv/config'
```
Comment (omar-mellouli) [inline src/server.ts]: always needed ?

File: src/services/messenger.ts [hunk 18]
```diff
@@ -65,50 +78,63 @@ class MessengerService {
             } else {
                 console.error(`Unknown Entry ${JSON.stringify(entry)}`)
             }
-        } catch (error) {
-            console.error(`Error when handling entry ${error}`)
+        } catch (error: unknown) {
+            console.error('Error when handling entry ', error)
```
Comment (omar-mellouli) [inline src/services/messenger.ts]: why remove use of Template strings ?

File: src/services/messenger.ts [hunk 19]
```diff
@@ -65,50 +78,63 @@ class MessengerService {
...
-            sendMessage(this.buildMessages(sender_psid, res))
-        } catch (error) {
-            console.error(`Error when handling message ${error}`)
+            console.log('Received message : ', received_message)
+			const boticot_data = await parseNlu(received_message, sender_psid)
+			const message = await this.buildMessages(sender_psid, boticot_data)
+			console.log('Built message : ', message)
+			sendMessage(message)
+        } catch (error: unknown) {
+            console.error('Error when handling message', error)
```
Comment (omar-mellouli) [inline src/services/messenger.ts]: why remove use of Template strings ?

File: src/services/messenger.ts [hunk 20]
```diff
@@ -65,50 +78,63 @@ class MessengerService {
...
 
-    async handleIntent(sender_psid, intent) {
+    async handleIntent(sender_psid: string, intent: string) {
         try {
-            const res = await requestIntent(intent, sender_psid)
-            sendMessage(this.buildMessages(sender_psid, res))
-        } catch (error) {
-            console.error(`Error when handling intent ${error}`)
+            const boticot_data = await requestIntent(intent, sender_psid)
+			const message = await this.buildMessages(sender_psid, boticot_data)
```
Comment (omar-mellouli) [inline src/services/messenger.ts:105]: remove extra space or rollback

File: src/client/nlu.ts (hunk 1 above)
Comment (leooel88) [reply src/client/nlu.ts]: DONE

File: package.json (hunk 2 above)
Comment (leooel88) [reply package.json]: DONE

File: typings/global.d.ts (hunk 5 above)
Comment (leooel88) [reply typings/global.d.ts]: DONE

File: typings/global.d.ts (hunk 6 above)
Comment (leooel88) [reply typings/global.d.ts]: DONE

File: typings/global.d.ts (hunk 7 above)
Comment (leooel88) [reply typings/global.d.ts]: DONE

File: src/utils/messenger.ts (hunk 10 above)
Comment (leooel88) [reply src/utils/messenger.ts]: DONE

File: src/services/messenger.ts (hunk 11 above)
Comment (leooel88) [reply src/services/messenger.ts]: DONE

File: src/services/messenger.ts (hunk 12 above)
Comment (leooel88) [reply src/services/messenger.ts]: DONE

File: typings/global.d.ts (hunk 13 above)
Comment (leooel88) [reply typings/global.d.ts]: DONE

File: typings/global.d.ts (hunk 14 above)
Comment (leooel88) [reply typings/global.d.ts]: DONE

File: environment.d.ts (hunk 15 above)
Comment (leooel88) [reply environment.d.ts]: DONE

File: package.json (hunk 16 above)
Comment (leooel88) [reply package.json]: DONE

File: src/routes/messenger.ts [hunk 21]
```diff
@@ -1,9 +1,14 @@
-import { Router } from 'express'
+import { Request, Response, Router } from 'express'
+import { WebhookVerificationRequest } from '../../typings/global'
 import messengerService from '../services/messenger'
 
 const router = Router();
 
-router.get('/webhook', (req, res): Promise<void> => messengerService.authorizeWebhook(req, res));
-router.post('/webhook', (req, res): Promise<void> => messengerService.handleWebhook(req, res));
+router.get('/webhook', (req: WebhookVerificationRequest, res: Response): void =>
```
Comment (omar-mellouli) [inline src/routes/messenger.ts:7]: Check if Ok when authorize webhook and when send message

File: src/services/messenger.ts [hunk 22]
```diff
@@ -65,50 +78,63 @@ class MessengerService {
...
+			sendMessage(message)
+        } catch (error: unknown) {
+            console.error('Error when handling intent', error)
```
Comment (omar-mellouli) [inline src/services/messenger.ts]: why remove use of Template strings ?

File: src/client/nlu.ts (hunk 3 above)
Comment (leooel88) [reply src/client/nlu.ts]: DONE

File: src/express.ts (hunk 4 above)
Comment (leooel88) [reply src/express.ts:7]: When pushing, tabs are added automaticaly
On local code, normal tabs (like master branch)

File: typings/express.d.ts (hunk 8 above)
Comment (leooel88) [reply typings/express.d.ts]: DONE

File: tsconfig.json (hunk 9 above)
Comment (leooel88) [reply tsconfig.json:14]: skipLibCheck needed

File: src/services/messenger.ts (hunk 20 above)
Comment (leooel88) [reply src/services/messenger.ts:105]: When pushing, tabs are added automaticaly
On local code, normal indentation

File: src/services/messenger.ts (hunk 19 above)
Comment (leooel88) [reply src/services/messenger.ts]: DONE : Template string back

File: src/services/messenger.ts (hunk 18 above)
Comment (leooel88) [reply src/services/messenger.ts]: DONE : Template string back

File: src/server.ts (hunk 17 above)
Comment (leooel88) [reply src/server.ts]: DONE : removed

**Review by omar-mellouli — CHANGES_REQUESTED** [review]

File: src/routes/messenger.ts (hunk 21 above)
Comment (leooel88) [reply src/routes/messenger.ts:7]: DONE : works with void

File: src/services/messenger.ts (hunk 22 above)
Comment (leooel88) [reply src/services/messenger.ts]: DONE : Template strings back

**Review by omar-mellouli — APPROVED** [review]

**Pull request merged**